File: validation/check_multipole.f90
!! Manual check that the multipole integrals are right
!!
!!     cmake -B build -DMQC_ENABLE_LIBCINT=ON
!!     ./build/check_multipole && python3 validation/check_multipole.py
!!
!! Milestone 1 of the EFP plan. Two things are being established, and they fail
!! differently:
!!
!!   * the integrals themselves, elementwise against `mol.intor('int1e_r')` and
!!     friends. A wrong component ordering or a wrong block offset shows up here
!!     and essentially nowhere else -- a transposed quadrupole still contracts to
!!     a plausible number against a density.
!!   * the origin actually reaching libcint. It travels through `env`, whose slot
!!     constants are 0-based and unconverted, and the failure mode is silent: the
!!     moments come back about the wrong point. The check for it is a *shifted*
!!     origin, because about the default origin a wrong slot gives the right
!!     answer by accident.
!!
!! The dipole from `Tr(D r)` is also printed. That number does not currently
!! exist on the CPU path at all -- only the GPU backend computes a dipole -- so
!! this is the first time it can be compared with anything.
program check_multipole
   use pic_types, only: dp
   use mqc_libcint_integrals, only: libcint_molecule_t, build_libcint_molecule
   use mqc_libcint_rhf, only: rhf_result_t, run_libcint_rhf
   use mqc_libcint_multipole, only: multipole_matrices
   use mqc_error, only: error_t
   implicit none

   real(dp), parameter :: ANG = 1.8897261254578281_dp
   integer :: failures

   failures = 0
   call one_case(1, "sto-3g", [0.0_dp, 0.0_dp, 0.0_dp])
   call one_case(2, "cc-pvdz", [0.0_dp, 0.0_dp, 0.0_dp])
   ! A shifted origin, which is the only arrangement that can catch the origin
   ! never arriving: at (0,0,0) a dropped origin is indistinguishable from a
   ! delivered one.
   call one_case(3, "cc-pvdz", [1.3_dp, -0.7_dp, 2.1_dp])

   write (*, "(A)") ""
   if (failures == 0) then
      write (*, "(A)") "[multipole] wrote every case; run validation/check_multipole.py"
   else
      write (*, "(A,I0,A)") "[multipole] ", failures, " FAILURE(S)"
      error stop 1
   end if

contains

   subroutine one_case(case_index, basis, origin)
      !! Files are named by index, not by the origin. Writing a float into a
      !! filename invites exactly the mismatch it caused: `F0.1` drops the
      !! leading zero, so 0.0 became ".0" and the Python side looked for a file
      !! that was never written. The basis and origin go *inside* the file.
      integer, intent(in) :: case_index
      character(len=*), intent(in) :: basis
      real(dp), intent(in) :: origin(3)

      type(libcint_molecule_t) :: mol
      type(rhf_result_t) :: scf
      type(error_t) :: err
      real(dp), allocatable :: dip(:, :, :), quad(:, :, :), oct(:, :, :)
      real(dp) :: c(3, 3), mu(3), nuc(3)
      integer :: unit, k, iatom
      character(len=64) :: name

      ! Water, off-centre and unsymmetric, so no component is zero by accident.
      c = reshape([0.0_dp, 0.0_dp, 0.0_dp, &
                   0.0_dp, 0.0_dp, 0.9584_dp*ANG, &
                   0.9268_dp*ANG, 0.0_dp, -0.2400_dp*ANG], [3, 3])

      call build_libcint_molecule([8, 1, 1], ["O ", "H ", "H "], c, basis, mol, err)
      if (err%has_error()) then
         write (*, "(A,A)") "[multipole] basis failed: ", err%get_message()
         failures = failures + 1
         return
      end if

      call multipole_matrices(mol, origin, 1, dip, err)
      if (.not. err%has_error()) call multipole_matrices(mol, origin, 2, quad, err)
      if (.not. err%has_error()) call multipole_matrices(mol, origin, 3, oct, err)
      if (err%has_error()) then
         write (*, "(A,A)") "[multipole] integrals failed: ", err%get_message()
         failures = failures + 1
         call mol%destroy()
         return
      end if

      call run_libcint_rhf(mol, 10, 100, 1.0e-10_dp, 1.0e-8_dp, .false., scf, err, &
                           in_core=.true.)
      if (err%has_error() .or. .not. scf%converged) then
         write (*, "(A)") "[multipole] SCF failed"
         failures = failures + 1
         call mol%destroy()
         return
      end if

      ! The electronic dipole is -Tr(D r); the nuclei contribute +sum Z (R - O).
      do k = 1, 3
         mu(k) = -sum(scf%density*dip(:, :, k))
      end do
      nuc = 0.0_dp
      do iatom = 1, mol%natm
         do k = 1, 3
            nuc(k) = nuc(k) + mol%charges(iatom)*(mol%coords(k, iatom) - origin(k))
         end do
      end do

      write (name, "(A,I0,A)") "/tmp/mqc_multipole_", case_index, ".txt"
      open (newunit=unit, file=trim(name), status="replace", action="write")
      write (unit, "(A)") trim(basis)
      write (unit, "(I0)") mol%nao
      write (unit, "(3es25.16e3)") origin
      write (unit, "(3es25.16e3)") mu + nuc
      write (unit, "(es25.16e3)") scf%energy
      ! The matrices themselves, for the elementwise comparison.
      write (unit, "(es25.16e3)") dip
      write (unit, "(es25.16e3)") quad
      write (unit, "(es25.16e3)") oct
      ! The overlap too. A basis function that is not exactly normalised is
      ! invisible to the SCF energy -- scaling a function does not change the
      ! space it spans, so the coefficients absorb it -- but it scales every
      ! per-AO quantity, including these moments. The overlap diagonal is where
      ! it shows.
      block
         real(dp), allocatable :: ovl(:, :)
         call mol%overlap(ovl)
         write (unit, "(es25.16e3)") ovl
         deallocate (ovl)
      end block
      close (unit)

      write (*, "(A,A,A,3F7.2,A,I0,A,3F14.8)") "  ", trim(basis), "  origin", origin, &
         "  nao ", mol%nao, "   dipole", mu + nuc
      call mol%destroy()
      deallocate (dip, quad, oct)
   end subroutine one_case

end program check_multipole

File: validation/check_multipole.py
#!/usr/bin/env python3
"""Compare our multipole integrals against PySCF's.

    ./build/check_multipole && python3 validation/check_multipole.py

Three questions, and they fail independently:

1. **Are the integrals right, elementwise?** Against ``mol.intor('int1e_r')``,
   ``int1e_rr`` and ``int1e_rrr`` on the same basis and the same origin. This is
   the check that catches a wrong component order or a wrong block offset -- a
   transposed quadrupole still contracts to a believable number against a
   density, so nothing downstream would object.

2. **Does the origin actually reach libcint?** It travels through ``env``, whose
   slot constants are 0-based and are not converted by the Fortran interface, so
   an off-by-one puts it in a neighbouring slot and the moments come back about
   the wrong point. The shifted-origin case is what settles it: about (0,0,0) a
   dropped origin is indistinguishable from a delivered one, and for a *neutral*
   molecule even the dipole is invariant either way. The quadrupole is not.

3. **Is the dipole right?** ``-Tr(D r)`` plus the nuclear term, against PySCF's
   own. Worth stating that this number did not previously exist on the CPU path
   at all -- only the GPU backend computed a dipole -- so this is the first
   opportunity to check it against anything.
"""

import sys
import pathlib

import numpy as np

REPO = pathlib.Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO / "tools" / "cpu_validation"))
from gen_cpu_validation import bse_to_pyscf, molecule_form, CARTESIAN  # noqa: E402

#: **Do not add a Pople basis with sp shells to this comparison without permuting.**
#: PySCF orders a Cartesian 6-31G* oxygen as `1s 2s 3s 2px..2pz 3px..3pz d`, grouping
#: all s functions then all p. Our reader hands libcint the s and p of each shared-
#: exponent group adjacently, giving `1s 2s 2px..2pz 3s 3px..3pz d`, which is also
#: GAMESS's order. So an elementwise comparison of AO-indexed matrices against PySCF
#: is only valid for a basis without sp shells -- the cases here are all like that,
#: which is why they pass.
#:
#: This was found the hard way: using PySCF's overlap with our own localized orbitals
#: reported them as non-orthonormal (`c^T S c` = 0.97, 0.75, 0.74, 0.43) when
#: `check_localize` verifies orthonormality to 2e-15 against our own overlap. The
#: metric was in the wrong order, not the orbitals.

ATOMS = [("O", (0.0, 0.0, 0.0)),
         ("H", (0.0, 0.0, 0.9584)),
         ("H", (0.9268, 0.0, -0.2400))]

CASES = [("sto-3g", (0.0, 0.0, 0.0)),
         ("cc-pvdz", (0.0, 0.0, 0.0)),
         ("cc-pvdz", (1.3, -0.7, 2.1))]

# Relative, because the octopole's matrix elements are an order of magnitude
# larger than the dipole's and an absolute bound would be a different test for
# each order.
#
# The floor is not our arithmetic. It is the basis files: BSE publishes exponents
# and coefficients to ten significant figures while PySCF's own tables carry
# more, which is worth ~1e-9 relative on any integral and is documented in
# gen_cpu_validation as looking exactly like a bug in whichever code is checked.
# Both sides here read the *same* basis_sets/ JSON, so what is left is that
# rounding propagating differently through two contraction routines.
INTEGRAL_TOL = 1e-8
DIPOLE_TOL = 1e-7


def read_dump(path):
    tokens = path.read_text().split()
    basis = tokens[0]
    values = [float(x) for x in tokens[1:]]
    nao = int(values[0])
    at = 1
    origin = np.array(values[at:at + 3]); at += 3
    dipole = np.array(values[at:at + 3]); at += 3
    energy = values[at]; at += 1
    out = {}
    for name, ncomp in (("dip", 3), ("quad", 9), ("oct", 27)):
        n = nao * nao * ncomp
        # Fortran order: (nao, nao, ncomp)
        block = np.array(values[at:at + n]).reshape((nao, nao, ncomp), order="F")
        at += n
        # to PySCF's (ncomp, nao, nao)
        out[name] = np.transpose(block, (2, 0, 1))
    out["overlap"] = np.array(values[at:at + nao * nao]).reshape((nao, nao), order="F")
    return dict(basis=basis, nao=nao, origin=origin, dipole=dipole,
                energy=energy, **out)


def main():
    from pyscf import gto, scf

    failures = 0
    for index, (_basis, _origin) in enumerate(CASES, start=1):
        path = pathlib.Path(f"/tmp/mqc_multipole_{index}.txt")
        if not path.exists():
            print(f"  MISSING {path} -- run ./build/check_multipole first")
            failures += 1
            continue
        ours = read_dump(path)
        basis, origin = ours["basis"], tuple(ours["origin"])

        symbols = {a[0] for a in ATOMS}
        mol = gto.Mole()
        mol.atom = ATOMS
        mol.unit = "Angstrom"
        mol.basis = {s: bse_to_pyscf(basis, s) for s in symbols}
        mol.cart = molecule_form(basis, symbols) == CARTESIAN
        mol.verbose = 0
        mol.build()
        mol.set_common_orig(origin)

        line = f"  {basis:8s} origin ({origin[0]:4.1f},{origin[1]:5.1f},{origin[2]:4.1f})"
        ok = True
        for name, intor in (("dip", "int1e_r"), ("quad", "int1e_rr"),
                            ("oct", "int1e_rrr")):
            theirs = mol.intor(intor)
            scale = max(np.max(np.abs(theirs)), 1.0)
            worst = np.max(np.abs(ours[name] - theirs))/scale
            line += f"  {name} {worst:.1e}"
            ok = ok and worst < INTEGRAL_TOL

        mf = scf.RHF(mol)
        mf.conv_tol = 1e-11
        mf.kernel()
        their_dipole = mf.dip_moment(unit="AU", verbose=0)
        d = float(np.max(np.abs(ours["dipole"] - their_dipole))
                  / max(float(np.max(np.abs(their_dipole))), 1.0))
        line += f"  dipole {d:.1e}"
        ok = ok and d < DIPOLE_TOL

        print(("  ok  " if ok else "  FAIL") + line)
        if not ok:
            failures += 1

    # The origin has to *matter*. If a shifted origin leaves the quadrupole
    # unchanged, it never reached libcint -- and neither the elementwise check
    # above nor any dipole would have said so, because PySCF would be asked with
    # the same origin and a neutral molecule's dipole is invariant regardless.
    a = pathlib.Path("/tmp/mqc_multipole_2.txt")
    b = pathlib.Path("/tmp/mqc_multipole_3.txt")
    if a.exists() and b.exists():
        qa, qb = read_dump(a)["quad"], read_dump(b)["quad"]
        moved = float(np.max(np.abs(qa - qb)))
        if moved < 1.0:
            print(f"  FAIL the quadrupole barely moved when the origin did "
                  f"({moved:.1e}); the origin is not reaching libcint")
            failures += 1
        else:
            print(f"  ok   shifting the origin moves the quadrupole by {moved:.3f}, "
                  f"so it is being applied")

    print()
    if failures:
        print(f"[multipole] {failures} FAILURE(S)")
        return 1
    print("[multipole] integrals, origin handling and the dipole all match PySCF")
    return 0


if __name__ == "__main__":
    sys.exit(main())
